pylops/optimization/solver.py

```diff
@@ -2,10 +2,202 @@
+        )
+        print("-----------------------------------------------------------------\n")
+
+    def setup(self, x0=None, niter=None, tol=1e-4, show=False):
+        """Setup solver
+
+        Parameters
+        ----------
+        x0 : :obj:`np.ndarray`, optional
+            Initial guess
+        niter : :obj:`int`, optional
+            Number of iterations
+        tol : :obj:`float`, optional
+            Tolerance on residual norm
+        show : :obj:`bool`, optional
+            Display setup log
+
+        """
+        self.tol = tol
+        self.niter = niter
+        if show:
+            self._print_setup()
+
+        # initialize solver
+        self.x0 = x0
+        if self.x0 is None:
+            self.x = self.ncp.zeros(self.Op.shape[1], dtype=self.y.dtype)
+            self.r = self.y.copy()
+        else:
+            self.x = self.x0.copy()
+            self.r = self.y - self.Op.matvec(self.x)
+        self.c = self.r.copy()
+        self.kold = self.ncp.abs(self.r.dot(self.r.conj()))
+
+        # create variables to track the residual norm and iterations
+        self.cost = []
+        self.cost.append(np.sqrt(self.kold))
+        self.iiter = 0
+
+    def step(self, show=False):
+        """Run one step of solver
+
+        Parameters
+        ----------
+        show : :obj:`bool`, optional
+            Display iteration log
+
+        """
+        Opc = self.Op.matvec(self.c)
+        cOpc = self.ncp.abs(self.c.dot(Opc.conj()))
+        a = self.kold / cOpc
+        self.x += a * self.c
+        self.r -= a * Opc
+        k = self.ncp.abs(self.r.dot(self.r.conj()))
+        b = k / self.kold
+        self.c = self.r + b * self.c
+        self.kold = k
+        self.iiter += 1
+        self.cost.append(np.sqrt(self.kold))
+        if show:
+            self._print_step()
+
+    def finalize(self, show=False):
+        """Finalize solver
+
+        Parameters
+        ----------
+        show : :obj:`bool`, optional
+            Display finalize log
+
+        """
+        if show:
+            self._print_finalize()
+        self.cost = np.array(self.cost)
+
+    def callback(self):
+        """Callback routine
+
+        This routine must be passed by the user as follows (when using the `solve`
+        method it will be automatically invoked after each step of the solve)
+
+        Examples
+        --------
+        >>> import numpy as np
+        >>> from pylops.basicoperators import Identity
+        >>> from pylops.optimization.solver import CG
+        >>> def callback():
+        ...     print('Running callback')
+        ...
+        >>> I = Identity(10)
+        >>> I
+        <10x10 Identity with dtype=float64>
+        >>> cgsolve = CG(I, np.ones(10))
+        >>> cgsolve.callback = callback
+
+        >>> cgsolve.callback()
+        Running callback
+        """
+        pass
+
+    def solve(self, x0=None, niter=10, tol=1e-4, show=False):

```

Comment:
Also move to baseclass? Most of the solver specific stuff is in `self.step()`. Again I suggest having a different interface,
```
x = self._step(x, *args)
self.callback(x)
```

Reply:
I would like this method to return as before. I don't see why you would call solve and then manually pick out the parts yourself is a better design choice.

Reply:
Sorry, this was a mistake. `callback` should be removed from CG and just put in base class.

Reply:
Mmmh I see your point but again I am not convinced I agree. Now we have an `x` that we keep passing in and out of the step...

I personally prefer the idea of relying on self.x, @cako we need your opinion here ;)

Reply:
Sorry, this was a mistake. callback should be removed from CG and just put in base class.

Reply:
Regarding solve returning values, this makes sense. I can do that in the next commit :)

Reply:
I don't love the self.x. Just make sure to set it to None before you return or something. So that the value is not left and someone forgets to call `setup()` and end up in weird situations. Of course these are user error, but a well-designed framework should be built to capture these things. Such issues could not arise if you keep `x` local.

Reply:
Ok, for this I think I would like a second opinion (@cako). 

If the problem is only `x`, I agree we can keep this out of the properties of the solver (see my last push if that makes sense, where I never even set self.x so I don't need to do the ugly thing of `self.x=None`... but if the problem is that you would like not to rely on properties at all, for Op, y, residual, hyperparams, etc. (which I hope is not what you mean), then I worry we are doing the Java mistake here, where we want to do something functional programming is good at but we do it with classes...

Reply:
I only have an issue with x, not the rest.

Reply:
Oook good, then it makes sense to me :)